torch/csrc/jit/tensorexpr/ir_simplifier.h

```diff
@@ -9,17 +11,260 @@ namespace torch {
+    }
+    t = promoteTypes(t, e.second->dtype());
+  }
+  return t;
+}
+
+template <class ExprType>
+Dtype promoteTypesVar(const ExprType* e) {
+  return e->dtype();
+}
+
+template <class ExprType, class... Args>
+Dtype promoteTypesVar(const ExprType* e, Args... es) {
+  Dtype lhs = e->dtype();
+  Dtype rhs = promoteTypesVar(es...);
+  if (e->isConstant()) {
+    lhs = Dtype(lhs.scalar_type(), rhs.lanes());
+  }
+
+  return promoteTypes(lhs, rhs);
+}
+
+// Helper for determining if an Expr is a multi-lane primitive (e.g. Broadcast
+// or Ramp).
+bool isMultilanePrimitive(const Expr* e) {
+  return e->expr_type() == IRNodeType::kBroadcast ||
+      e->expr_type() == IRNodeType::kRamp;
+}
+} // namespace
+
+// A Term represents a grouping of Exprs through multiplication.
+// E.g. product(scalar, *variables).
+class Term : public ExprNode<Term> {
+ public:
+  template <class... Args>
+  Term(HashProvider& hasher, const Expr* s, Args... ts)
+      : ExprNodeBase(promoteTypesVar(s, ts...)), scalar_(s), hasher_(hasher) {
+    CHECK(s->isConstant());
+    addComponent(ts...);
+    sort();
+  }
+
+  Term(HashProvider& hasher, const Expr* s, std::vector<const Expr*> v)
+      : ExprNodeBase(promoteTypesVec(s, v)),
+        variables_(std::move(v)),
+        scalar_(s),
+        hasher_(hasher) {
+    sort();
+  }
+
+  // Convenience constructor from a map of hash -> var, used when merging Terms.
+  Term(
+      HashProvider& hasher,
+      const Expr* s,
+      std::unordered_map<SimplifierHashType, const Expr*> varmap)
+      : ExprNodeBase(promoteTypesMap(s, varmap)), scalar_(s), hasher_(hasher) {
+    for (auto& p : varmap) {
+      addComponent(p.second);
+    }
+    sort();
+  }
+
+  const Expr* scalar() const {
+    return scalar_;
+  }
+  const std::vector<const Expr*>& variables() const {
+    return variables_;
+  }
+  HashProvider& hasher() const {
+    return hasher_;
+  }
+
+  // Produce a hash of just the variable components of this term, to determine
+  // if it can be combined with another term.
+  SimplifierHashType hashVars() const {
+    SimplifierHashType hash = 0;
+    for (auto* v : variables_) {
+      hash = hasher_.hash_combine(hash, v);
+    }
+    return hash;
+  }
+
+ private:
+  std::vector<const Expr*> variables_;
+  const Expr* scalar_;
+  HashProvider& hasher_;
+
+  void addComponent() {}
+  void addComponent(const Expr* e) {
+    variables_.push_back(e);
+  }
+  template <class... Es>
+  void addComponent(const Expr* e, Es... es) {
+    addComponent(e);
+    addComponent(es...);
+  }
+
+  // Sort by hash.
+  void sort() {
+    // order of ops important for float
+    if (dtype().is_floating_point()) {
+      return;
+    }
+    std::sort(
+        variables_.begin(),
+        variables_.end(),
+        [&](const Expr* a, const Expr* b) {
+          return hasher_.hash(a) < hasher_.hash(b);
+        });
+  }
+};
+
+// Polynomial represents a grouping of Exprs by addition.
+// E.g. sum(*variables, scalar).
```

Comment:
What is scalar here? Can the component be something else (neither Var nor FloatImm/IntImm)?

Reply:
Scalar term must be Immediate, variable terms can be anything (Var, compound Expr of any ops, Polynomial or Term).